### ISSUE TITLE
[Feature] Hook code sharing

### ISSUE THREAD

**Issue by MatthewScholefield** [Escape-Technologies/mookme#25]
At the moment I'm working on a project with 3 Python packages (but soon to be 5+) and each package has a nearly identical `.hooks/pre-commit.json`. I'm opening this issue to discuss the best approach towards sharing pre-commit files and handling shared logic. For reference, here is the pre-commit for one of the packages:
```json
{
  "type": "python",
  "venvActivate": ".venv/bin/activate",
  "steps": [
    {
      "name": "Format Code",
      "command": "blue .",
      "onlyOn": "**/*.py",
      "serial": true
    },
    {
      "name": "Sort Imports",
      "command": "isort . --profile attrs",
      "onlyOn": "**/*.py",
      "serial": true
    },
    {
      "name": "Ensure Style (flake8)",
      "command": "flake8 my_package",
      "onlyOn": "**/*.py"
    },
    {
      "name": "Ensure Style (pylint)",
      "command": "git diff --cached --name-only --relative | while read line; do [ ! -f \"$line\" ] || printf \"%s\\n\" \"$line\"; done | grep -P '\\.py$' | xargs -d '\\n' pylint -E --extension-pkg-whitelist='pydantic' --ignored-modules=websockets",
      "onlyOn": "**/*.py"
    }
  ]
}
```

So, there are two problems I think would be nice to handle:
 - Sharing hooks for multiple similar packages (ie. between all Python packages)
 - Sharing bash code (like `git diff --cached --name-only ...`)

My current thought is that I can make some `/.hooks/base/python.json` which executes, for example, `flake8 $(../.hooks/utils/vars .package)` with some `/.hooks/utils/vars` script like:
```bash
#!/usr/bin/env bash
cat .hooks/vars.json | jq -r "$1"
```

Finally, I could just make `/my-package/.hooks/pre-commit.sh` a link to `/.hooks/base/python.json` to have a single source of truth for all Python package pre-commit hooks.

Then, I could also have some other utility files like `/.hooks/utils/changed-files`:
```bash
#!/usr/bin/env bash
git diff --cached --name-only --relative | while read line; do [ ! -f \"$line\" ] || printf \"%s\\n\" \"$line\"; done | grep -P '\\.py$'
```

Is this a reasonable approach? If so, what about supporting injecting `/.hooks/utils` into `$PATH` so that we could then write things like `flake8 $(vars .package)` and `changed-files | xargs ...`?

Also, this might be going a little crazy, but one option in general (that works whether or not the above is performed) is to write hooks as `.jsonnet` files and then use a pre-commit hook to generate the actual hook json files.

**Comment by LMaxence** [comment]
Hello ! First of all, I'm truly sorry for the delay, I just came back from holidays.

Just to make sure I wrap my head correctly around this one, you would like to use some sort of *partials* for writing hooks, and the second suggestion is about writing templated hooks, and provide them with variables on a per-package basis right ?

I really like these ideas and I'm eager to work on this, but this will be a piece of work, and I might need your help for designing use-cases and needs. 

**One thing that would really help me understand this correctly, would be that you provide us with an example of resulting `mookme` hooks file (global and package files) implementing such features.**

For instance, what I have in mind is the following result:

```sh
# /.hooks/_partials/flake8.json
{
  "name": "Ensure Style (flake8)",
  "command": "flake8 {{PACKAGE_NAME}}",
  "onlyOn": "**/*.py"
}
```

```sh
# /some_package/.hooks/flake8.json
{
  "extends": "flake8",
  "variables": {
    "PACKAGE_NAME":  "some_package"
  }
}
```

Am i missing something ?

**Comment by LMaxence** [comment]
I'm reading this again and I figured out I misunderstood your point. What you want to be reusable are scripts right ? So you would like to be able to define commands in the global hooks folder and invoke them in the packages' hook files ?

**Comment by MatthewScholefield** [comment]
Yeah, so what you suggested would work but I was thinking maybe just inserting some utils path into PATH would be a simpler more general solution. Maybe that combined with automatically reading some `.hooks/env` file or something where we could define environment variables for the scripts.

Then a simple symlinked script would allow the entore hooks file to be shared.

**Comment by LMaxence** [comment]
Hello Matthew,

Just to give you visibility about this, I will have a window on Friday of this week to develop a poc of this. Would be great to get a review from you to make sure I'm not missing your point :)

I really love this idea

**Comment by LMaxence** [comment]
Hello again here, as I start working on this, I realize that I might not have fully understood this part:

> Then, I could just make /my-package/.hooks/pre-commit.sh a link to /.hooks/base/python.json to have a single source of truth for all Python package pre-commit hooks.

It seems a bit tricky to change the package hook files from `json` to shell script, since the `json` structure holds some useful metadata that a shell script file could not hold.

Waiting for some clarification on this, I'm gonna implement the rest of this feature in this order:

1. Hooks sharing: You will be able to declare base hooks in your hook root folder. This means that for instance, this structure
```
project-root
 |------- .hooks
     |------- shared
            |------- flake8.json
 |------- package.json
 |   |------- packages
 |         |------- some-package
                 |------- .hooks
                       |------- pre-commit.json
```
will give you the ability to re-use the flake8 shared hook in `some-package/.hooks/pre-commit.json`

2. Hooks partial scripts: I'll add another folder along the `shared` folder in the `.hooks` root directory, something like `partial` in which you will be able to declare shell scripts, that will be added to the `$PATH` variable at startup giving you the ability to do the `flake8 $(vars .package)` and `$(changed-files)` stuff

**Comment by LMaxence** [comment]
@MatthewScholefield I'll wait for your review to make sure it matches what you had in mind :)

**Comment by LMaxence** [comment]
@MatthewScholefield I will merge this feature today or tomorrow, if you have feedbacks to give me now is the time. I'll publish the new version featuring these changes at the same time :)
